### PR TITLE
109 agenda session registrations

### LINKED ISSUE #109: Add Agenda + Session registration
Migrate the equivalent page from the previous app (found on the `react-old` branch) to Angular.

Add both the list page (i.e. `Venues Page`) as well as the Individual page (i.e. `Venue Page`) and any extra components used by it on the old version if they haven't been added yet.

TIP: try using ChatGPT to translate the code from React to Angular -> it won't be perfect but it's a good starting base

### LINKED ISSUE #111: Add a menu to create and manage entities
Copy logic from old React app
Also, change entity managing/creation permission in front and back-end from `isAdmin` to `canManageContent`

Migrate the equivalent page from the previous app (found on the `react-old` branch) to Angular.

Add both the list page (i.e. `Venues Page`) as well as the Individual page (i.e. `Venue Page`) and any extra components used by it on the old version if they haven't been added yet.

TIP: try using ChatGPT to translate the code from React to Angular -> it won't be perfect but it's a good starting base

### PR DESCRIPTION
**Pull request by rbento1096**
closes #109 
closes #111 
relates to #78 

Once merged we will do some graphical improvements and bug fixing if required and then we can merge into prod.

things to improve (this PR or a new issue):
- some @todos in the sessions page (i.e. we need to open the detail in a modal for mobile) + when pressing a button it automatically selects the session (ev.preventDefault() not working for some reason...). + other minor things.
- Some graphical improvements can be made to most lists (both entity lists as well as the lists inside entity details).
   - example: Organizations page could use a grid with ion-card dispalying the company logo. Same for Speakers
   - example: Room detail page has a list with ion-items with too much space, we can reduce margins and add more content and color to the list (session code, etc..). Also check if the links work, there is no `session/sessionId` page
 
- to ask:
   - should we hide past sessions? maybe on toggle?
   - can users favorite sessions without registering?
   - should the session detail page include a list of the participants in the session (it has the needed data for that)

### CODE REVIEW

File: back-end/src/handlers/registrations.ts [hunk 1]
```diff
@@ -7,6 +7,7 @@ import { DynamoDB, HandledError, ResourceController } from 'idea-aws';
 import { Session } from '../models/session.model';
```
Comment (uatisdeproblem) [inline back-end/src/handlers/registrations.ts:7]: suggestion: I'd rename this RC "sessionsRegistrations", not to be confused with users registrations

File: back-end/src/handlers/registrations.ts [hunk 2]
```diff
@@ -72,13 +86,15 @@ class SessionRegistrations extends ResourceController {
...
+    if (!this.configurations.areSessionRegistrationsOpen) throw new HandledError('Registrations are closed!')
 
     this.registration = new SessionRegistration({
       sessionId: this.resourceId,
-      userId: this.principalId,
-      registrationDateInMs: new Date().getTime()
+      userId: this.user.userId,
+      registrationDateInMs: new Date().getTime(),
+      name: this.user.getName(),
+      esnCountry: this.user.sectionCountry
```
Comment (uatisdeproblem) [inline back-end/src/handlers/registrations.ts:97]: would make sense to keep `sectionCountry`

File: back-end/src/handlers/sessions.ts [hunk 3]
```diff
@@ -69,20 +69,20 @@ class Sessions extends ResourceController {
...
     );
 
-    this.session.speakers = (
-      await ddb.batchGet(
-        DDB_TABLES.speakers,
-        this.session.speakers?.map(speakerId => ({ speakerId })),
-        true
-      )
-    ).map(s => new SpeakerLinked(s));
+    const getRooms = await ddb.batchGet(
```
Comment (uatisdeproblem) [inline back-end/src/handlers/sessions.ts:76]: getSpeakers

File: back-end/src/models/configurations.model.ts [hunk 4]
```diff
@@ -6,6 +6,8 @@ import { ServiceLanguages } from './serviceLanguages.enum';
 
 export const LANGUAGES = new Languages({ default: ServiceLanguages.English, available: [ServiceLanguages.English] });
 
+const DEFAULT_SESSION_REGISTRATION_BUFFER = 10;
```
Comment (uatisdeproblem) [inline back-end/src/models/configurations.model.ts:9]:
```suggestion
const DEFAULT_SESSION_REGISTRATION_BUFFER_MINUTES = 10;
```

File: back-end/src/models/session.model.ts [hunk 5]
```diff
@@ -92,58 +94,43 @@ export class Session extends Resource {
...
   }
 
   // @todo add a method to check if a user/speaker is in the session or not
 
-  calcDatetimeWithoutTimezone(dateToFormat: Date | string | number): datetime {
+  calcDatetimeWithoutTimezone(dateToFormat: Date | string | number, bufferInMinutes = 0): datetime {
     const date = new Date(dateToFormat);
-    return new Date(date.getTime() - date.getTimezoneOffset() * 60000).toISOString().slice(0, 16);
+    return new Date(
+      date.getTime() -
```
Comment (uatisdeproblem) [inline back-end/src/models/session.model.ts:106]: After the merge I can come back to this timezone logic and simplify it, similarly to how we do in the GA app

File: back-end/src/models/sessionRegistration.model.ts [hunk 6]
```diff
@@ -13,11 +13,21 @@ export class SessionRegistration extends Resource {
...
    */
   registrationDateInMs: number;
+  /**
+   * The user's name.
+   */
+  name: string;
+  /**
+   * The user's ESN Country if any.
+   */
+  esnCountry?: string;
```
Comment (uatisdeproblem) [inline back-end/src/models/sessionRegistration.model.ts:23]: if it's optional below it should be if(x.esnCountry); also, I would use the same naming we previously used: sectionCountry

File: back-end/src/models/speaker.model.ts [hunk 7]
```diff
@@ -50,6 +55,10 @@ export class Speaker extends Resource {
       typeof x.organization === 'string'
         ? new OrganizationLinked({ organizationId: x.organization })
         : new OrganizationLinked(x.organization);
+    this.socialMedia = {}
```
Comment (uatisdeproblem) [inline back-end/src/models/speaker.model.ts:58]: After the merge I can come back with a better structure for social media, if you want (copying Papaya)

File: back-end/src/models/user.model.ts [hunk 8]
```diff
@@ -95,6 +99,10 @@ export class User extends Resource {
     this.registrationForm = x.registrationForm ?? {};
     if (x.registrationAt) this.registrationAt = this.clean(x.registrationAt, t => new Date(t).toISOString());
     if (x.spot) this.spot = new EventSpotAttached(x.spot);
+    this.socialMedia = {}
```
Comment (uatisdeproblem) [inline back-end/src/models/user.model.ts:102]: After the merge I can come back with a better structure for social media, if you want (copying Papaya)

File: back-end/src/models/speaker.model.ts [hunk 9]
```diff
@@ -58,7 +67,6 @@ export class Speaker extends Resource {
   validate(): string[] {
     const e = super.validate();
     if (isEmpty(this.name)) e.push('name');
-    if (this.contactEmail && isEmpty(this.contactEmail, 'email')) e.push('contactEmail');
```
Comment (uatisdeproblem) [inline back-end/src/models/speaker.model.ts:61]: I think a contact email would be nice to have for a speaker, right?

File: back-end/src/models/speaker.model.ts (hunk 7 above)
Comment (uatisdeproblem) [reply back-end/src/models/speaker.model.ts:58]: It could be part of socialMedia

File: front-end/src/app/app.service.ts [hunk 10]
```diff
@@ -208,4 +208,13 @@ export class AppService {
     const p = this.user.permissions;
     return p.isAdmin || p.canManageRegistrations || p.canManageContents;
   }
+
+  formatDateShort = (date: string | Date): string => {
```
Comment (uatisdeproblem) [inline front-end/src/app/app.service.ts:212]: i'd suggest using pipes/have a logic more similar to GA app; I can come back to this if you want

**Review by uatisdeproblem — APPROVED** [review]
Generally, if you find it useful, I can go back to refactor some code (as suggested in the comments) for more stability and clarity, between Wednesday and Sunday. Or I can work on other stuff, as you prefer :)

File: back-end/src/models/speaker.model.ts (hunk 9 above)
Comment (rbento1096) [reply back-end/src/models/speaker.model.ts:61]: I agree, but this year's data didn't include one so I didn't make this mandatory :)

**Comment by rbento1096** [comment]
@uatisdeproblem from your review I have resolved those that have been fixed in recent commits. The other comments I agree with them but maybe it is better to make them after the event so as not to risk breaking anything. At least those related to data.

I created the issue #126 to keep track of these. If instead, you feel that there is no risk and they are easily doable in the next issues (the pipes for example). feel free to do it, just let me know so I can resolve the comments here :)

Thank you a lot for your help!

**Pull request merged**